menu.py

```diff
@@ -1,38 +1,452 @@
+
+VK_NAMES = {
+    kVK_Return: "enter",
+    kVK_Tab: "tab",
+    kVK_Space: "space",
+    kVK_Delete: "backspace",
+    kVK_Escape: "esc",
+    # kVK_Command: '',
+    # kVK_Shift: '',
+    # kVK_CapsLock: '',
+    # kVK_Option: '',
+    # kVK_Control: '',
+    # kVK_RightCommand: '',
+    # kVK_RightShift: '',
+    # kVK_RightOption: '',
+    # kVK_RightControl: '',
+    # kVK_Function: '',
+    # kVK_F17: '',
+    # kVK_VolumeUp: '',
+    # kVK_VolumeDown: '',
+    # kVK_Mute: '',
+    kVK_F18: "f18",
+    kVK_F19: "f19",
+    kVK_F20: "f20",
+    kVK_F5: "f5",
+    kVK_F6: "f6",
+    kVK_F7: "f7",
+    kVK_F3: "f3",
+    kVK_F8: "f8",
+    kVK_F9: "f9",
+    kVK_F11: "f11",
+    kVK_F13: "f13",
+    kVK_F16: "f16",
+    kVK_F14: "f14",
+    kVK_F10: "f10",
+    kVK_F12: "f12",
+    kVK_F15: "f15",
+    # kVK_Help: '',
+    kVK_Home: "home",
+    kVK_PageUp: "pageup",
+    kVK_ForwardDelete: "delete",
+    kVK_F4: "f4",
+    kVK_End: "end",
+    kVK_F2: "f2",
+    kVK_PageDown: "pagedown",
+    kVK_F1: "f1",
+    kVK_LeftArrow: "left",
+    kVK_RightArrow: "right",
+    kVK_DownArrow: "down",
+    kVK_UpArrow: "up",
+    # kVK_ISO_Section: '',
+    # kVK_JIS_Yen: '',
+    # kVK_JIS_Underscore: '',
+    # kVK_JIS_KeypadComma: '',
+    # kVK_JIS_Eisu: '',
+    # kVK_JIS_Kana: '',
+}
+
+
+def active_menu_bar():
     return ui.active_app().children.find_one(AXRole="AXMenuBar", max_depth=0)
 
 
+def selected_menu_and_path():
+    """Returns selected element in menu bar, and path to it"""
+    selected_menu = active_menu_bar()
+    menu_path = []
+    while True:
+        if not selected_menu.AXSelectedChildren:
+            break
+        selected_menu = selected_menu.AXSelectedChildren[0]
+
+        if not (title := selected_menu.get("AXTitle")):
+            return selected_menu, menu_path
+
+        menu_path.append(selected_menu.AXTitle)
+        # XXX(nriley) this is True even if list empty - report Talon bug? children also doesn't support len()
+        # print(selected_menu.children, bool(selected_menu.children))
+        try:
+            selected_menu = selected_menu.children.find_one(
+                AXRole="AXMenu", max_depth=0
+            )
+        except ui.UIErr:
+            break
+
+    return selected_menu, menu_path
+
+
+def mouse_pos_menu_and_path():
+    """Returns element in menu bar under cursor, and path to it"""
+    menu_path = []
+    element_at_mouse_pos = ui.element_at(*ctrl.mouse_pos())
+    if element_at_mouse_pos.AXRole in ("AXMenu", "AXMenuItem", "AXMenuBarItem"):
+        menu = element_at_mouse_pos
+        while True:
+            if menu.AXRole in ("AXMenuItem", "AXMenuBarItem"):
+                menu_path.append(menu.AXTitle)
+            menu = menu.AXParent
+            if menu.AXRole == "AXMenu":
+                menu = menu.AXParent
+            elif menu.AXRole == "AXMenuBar":
+                menu_path.reverse()
+                break
+        else:  # not in menu bar
+            return None, []
```

Comment:
PyCharm says this is unreachable. I think it needs to be indented one more level?

Reply:
Someday, `True` may be false, I swear. Thanks, fixed.